### PR TITLE
fix: set allocation start time accurately (specifically for Kubernetees) [DET-6713]

### PR DESCRIPTION
**Pull request by nrajanee**
## Description

fix: set allocation start time accurately (specifically for Kubernetees) [DET-6713]

## Test Plan

Tested in unit test master/internal/task/allocation_test.go. No manual test is required. 

[DET-6713]: https://determinedai.atlassian.net/browse/DET-6713?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: master/internal/task/allocation.go
Comment (stoksc) [inline master/internal/task/allocation.go:398]: we need to make sure this gets saved immediately, i'm not sure when the next time the model is sync is

File: master/internal/db/postgres_tasks.go
Comment (stoksc) [inline master/internal/db/postgres_tasks.go:116]: i'm not sure i understand these changes, and the others in this file. i think we should just be setting start time in allocation.go and saving it.

File: master/internal/task/allocation.go
Comment (nrajanee) [reply master/internal/task/allocation.go:398]: Got it! I see an UpdateAllocationState method in master/db/database.go but that only updates the allocation state. Should I create a similar method that updates the start time?

File: master/internal/db/postgres_tasks.go
Comment (nrajanee) [reply master/internal/db/postgres_tasks.go:116]: These changes are related to these concerns we had when we initially discussed the implementation: - CompleteAllocation if start time is zero value set to cluster_heartbeat. This can happen when we call cleanup after a Poststop message. This message is sent after panic or error. 
- CloseOpenAllocations if start time is zero value set to cluster_heartbeat.  This can happen when cluster fails. 

If we don't do this then we can end up with a zero value in start time and end_time as the current value or cluster_heartbeat which is incorrect.

File: master/internal/task/allocation.go
Comment (nrajanee) [reply master/internal/task/allocation.go]: done

File: master/internal/task/allocation.go
Comment (nrajanee) [reply master/internal/task/allocation.go:397]: done. Should I add a check for this to the postgres_task_intg_test.go?

File: master/internal/db/postgres_tasks.go
Comment (stoksc) [inline master/internal/db/postgres_tasks.go:105]: the diff between this and the former version is that we used to make sure w only affected 1 row. maybe leave that check in (nice to know it happens somewhere, even those it _should_ never be broken).

File: master/internal/db/postgres_tasks.go
Comment (stoksc) [inline master/internal/db/postgres_tasks.go:201]:
```suggestion
		SET start_time=$2
```

File: master/internal/db/postgres_tasks.go
Comment (stoksc) [inline master/internal/db/postgres_tasks.go:211]: nit, here and elsewhere format the queries consistently. also i would run these queries in a transaction.

File: master/internal/task/allocation.go
Comment (stoksc) [inline master/internal/task/allocation.go:398]: i think we can just log this error and move on. log an error that says something along the lines of 'this allocation maybe not be properly accounted for'.

File: master/internal/task/allocation_test.go
Comment (stoksc) [inline master/internal/task/allocation_test.go:114]: if you use a format string you need to use `require.Falsef` instead. and you can drop the `.String()` calls and just use `%s` i think.

File: master/internal/db/postgres_tasks.go
Comment (nrajanee) [reply master/internal/db/postgres_tasks.go:105]: But we're using allocation_id here so that means we'll only affect one row?
Leave the check in as in check for a.startTime == nil? Then'll I'll have to have two separate queries like before?

Comment (nrajanee) [reply master/internal/db/postgres_tasks.go:105]: > the diff between this and the former version is that we used to make sure w only affected 1 row. maybe leave that check in (nice to know it happens somewhere, even those it _should_ never be broken).

But we're using allocation_id here so that means we'll only affect one row?
Leave the check in as in check for a.startTime == nil? Then'll I'll have to have two separate queries like before?

File: master/internal/db/postgres_tasks.go
Comment (nrajanee) [reply master/internal/db/postgres_tasks.go:201]: Oops missed this! thank you.

File: master/internal/db/postgres_tasks.go
Comment (nrajanee) [reply master/internal/db/postgres_tasks.go:211]: will do!

File: master/internal/db/postgres_tasks.go
Comment (stoksc) [reply master/internal/db/postgres_tasks.go:105]: no i mean leave the check that we only affected one row.

> But we're using allocation_id here so that means we'll only affect one row?

true yeah, i was being extra cautious since id is generated by us but i think there is a unique constraint on id so it's ok.

File: master/internal/db/postgres_tasks.go
Comment (nrajanee) [reply master/internal/db/postgres_tasks.go:211]: locally, ran the queries in a transaction and they work fine!

**Review by stoksc — APPROVED** [review]
lgtm modulo final comments

**Comment by stoksc** [comment]
@nrajanee make sure to rebase and ensure your migration is last still before landing

File: master/static/migrations/20220315162743_drop-not-null-allocation-start-time.tx.down.sql
Comment (stoksc) [reply master/static/migrations/20220315162743_drop-not-null-allocation-start-time.tx.down.sql:3]: @nrajanee i think this got missed, fyi

**Pull request merged**
